### PR TITLE
#18 Extend action to remove org join to all users

### PR DESCRIPTION
**Pull request by nmadruga**
Based in: #874 (I think it was the one used for latest build...)
Fixes https://github.com/openmsupply/conforma-templates/issues/18

### Testing
- Load [disableOrg-3 template](https://github.com/openmsupply/conforma-server/files/9760354/disableOrg-3.zip) in latest snap shot. Make sure is set as Available.
- Try removing one organisation. (ex: 13) - Log in as `manager` and use the template (Top Menu => Gerência => Desativar Entidade)
- See how it's removed by running this Graphql query **before** and **after**

### Changes

1. Allow passing `username` and/or `orgName` instead of ides
2. Enforces that some orgId or orgName was passed - it cannot be null
3. When no userId or username is passed (`undefined`) will interpret as remove ALL used connected to org
4. Keeps same logic to remove permissionJoins - for the pair `{userId, orgId}` or for all users connected to org (when no userId is defined).

~~TODO: Update documentation~~

### CODE REVIEW

**Comment by CarlosNZ** [comment]
Looks good. Would you mind updating the reference docs (`List-of-Action-plugins.md`) to reflect the new API, it'll make it easier to review (as well as keeping it accurate)?

Thanks.

**Comment by nmadruga** [comment]
Done 6a303d0

**Comment by nmadruga** [comment]
I think it was using Prettier to format .md files too. Now I have installed this extension, but it does add this automatic sub-headers in the contents list. I'll see how to remove this settings
Screen Shot 2022-10-14 at 9 50 16 AM

**Comment by nmadruga** [comment]
> I think it was using Prettier to format .md files too. Now I have installed this extension, but it does add this automatic sub-headers in the contents list. I'll see how to remove this settings Screen Shot 2022-10-14 at 9 50 16 AM

Sorted ;)
Screen Shot 2022-10-14 at 9 55 16 AM

File: plugins/action_remove_user_org/src/removeUserOrg.ts
Comment (CarlosNZ) [inline plugins/action_remove_user_org/src/removeUserOrg.ts:48]: This has to be a named property, not an array. Cos the `output` from multiple actions is cumulatively built into the "cumulativeOutput" object by merging the output of each sequential action object. 

So this will break that.

Probably just do `output: {removedUsers: removed.pairs}`

Also -- can you have a think whether any of our current configs use the output of this action anywhere? Cos if so, those will need to be updated

File: plugins/types.ts
Comment (CarlosNZ) [inline plugins/types.ts:13]: Please revert, as per above comment.

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Looks good, thanks.

Just one little thing that will need changing. See comment for explanation.

File: plugins/action_remove_user_org/src/removeUserOrg.ts
Comment (CarlosNZ) [reply plugins/action_remove_user_org/src/removeUserOrg.ts:48]: Oh, and remember to update the docs accordingly :)

Comment (nmadruga) [reply plugins/action_remove_user_org/src/removeUserOrg.ts]: Oh, that makes a lot of sense! Thanks

No, I don't think this is used by any other action as an input

File: documentation/List-of-Action-plugins.md
Comment (CarlosNZ) [inline documentation/List-of-Action-plugins.md:414]:
```suggestion
| `username` or `userId`                     | `removedUsers: {userOrgId, userId, orgId} [Array]` |
```
Oops -- don't you mean "removedUsers"? ;)

**Review by CarlosNZ — APPROVED** [review]
Great, working nicely, output looks correct.

Can you just fix that little documentation thingy and then good to merge: https://github.com/openmsupply/conforma-server/pull/904#discussion_r995278502

👍

**Comment by nmadruga** [comment]
> Can you just fix that little documentation thingy and then good to merge: [#904 (comment)](https://github.com/openmsupply/conforma-server/pull/904#discussion_r995278502)

Wow - Wasn't I distracted on that day! Thanks! Fixed now, will continue with merge.

**Comment by nmadruga** [comment]
Oh... actually this is pointing to `feature/data-view-filter` branch, since we are using this as the base to create latests builds. I'll wait until that is merged to get this PR in also. I don't want to mess with the merge-history by re-basing this to `develop`

**Comment by CarlosNZ** [comment]
> Oh... actually this is pointing to `feature/data-view-filter` branch, since we are using this as the base to create latests builds. I'll wait until that is merged to get this PR in also. I don't want to mess with the merge-history by re-basing this to `develop`

Okay, they're pretty well separated so I think you'd be fine to rebase to develop. But I can also merge the filters branch today if you're happy with it. Maybe after the meeting?

**Pull request merged**
